File: sdk/python-langchain/tests/_httpd.py
"""Minimal HTTP server harness for `HtmlConnector`'s URL-fetch tests (issue
#349) — a stdlib `ThreadingHTTPServer` serving canned, in-memory responses,
synthesized at test time rather than depending on any external service or
recorded fixture, the same "no binary fixture" convention `tests/_pdfs.py`
already established for PDF bytes.
"""

from __future__ import annotations

import threading
import time
from collections.abc import Iterator
from contextlib import contextmanager
from dataclasses import dataclass, field
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from urllib.parse import urlsplit


@dataclass(frozen=True)
class Route:
    """One canned response. `location` set makes this a 302 redirect
    (`status`/`content_type`/`body` are then ignored); otherwise `status` is
    served with `body`, and `content_type` becomes the `Content-Type` header
    (omit to send no `Content-Type` header at all). `delay` (seconds) sleeps
    before responding at all — enough to exercise a caller's per-phase
    `timeout`. `chunk_delay` instead trickles `body` out in small pieces,
    sleeping between each — every individual gap stays under a per-phase
    `timeout`, so this is what exercises a caller's separate total-time
    budget instead; no `Content-Length` is sent for a trickled response
    (`protocol_version` defaults to HTTP/1.0, so the client reads until the
    connection closes)."""

    status: int = 200
    content_type: str | None = "text/html; charset=utf-8"
    body: bytes = b""
    delay: float = 0.0
    chunk_delay: float = 0.0
    location: str | None = None
    headers: dict[str, str] = field(default_factory=dict)


class _Handler(BaseHTTPRequestHandler):
    server: _RouteServer  # narrows the base class's plain socketserver type

    def log_message(self, format: str, *args: object) -> None:  # noqa: A002 - stdlib signature
        pass  # keep test output quiet; failures still show via assertions

    def do_GET(self) -> None:  # noqa: N802 - stdlib-mandated method name
        # `self.path` is the full request target, query string included —
        # route registration is by path only, so a request carrying a query
        # string (the credential/token-stripping tests deliberately send
        # one) must not be matched against it too.
        route = self.server.routes.get(urlsplit(self.path).path)
        if route is None:
            self.send_response(404)
            self.end_headers()
            return
        if route.delay:
            time.sleep(route.delay)
        if route.location is not None:
            self.send_response(302)
            self.send_header("Location", route.location)
            self.end_headers()
            return
        self.send_response(route.status)
        if route.content_type is not None:
            self.send_header("Content-Type", route.content_type)
        for name, value in route.headers.items():
            self.send_header(name, value)
        if route.chunk_delay:
            self.end_headers()
            chunk_size = 16
            for offset in range(0, len(route.body), chunk_size):
                self.wfile.write(route.body[offset : offset + chunk_size])
                self.wfile.flush()
                time.sleep(route.chunk_delay)
            return
        self.send_header("Content-Length", str(len(route.body)))
        self.end_headers()
        self.wfile.write(route.body)


class _RouteServer(ThreadingHTTPServer):
    def __init__(self, routes: dict[str, Route]) -> None:
        super().__init__(("127.0.0.1", 0), _Handler)
        self.routes = routes

    @property
    def base_url(self) -> str:
        return f"http://127.0.0.1:{self.server_address[1]}"

    def handle_error(self, request: object, client_address: object) -> None:
        # A client that gives up mid-response (the `timeout`/oversized-body
        # tests deliberately do this) leaves this server thread writing to a
        # closed socket; the base class's default `handle_error` prints a
        # full traceback to stderr for that, which is expected here, not a
        # real failure — swallow only the connection-teardown exceptions,
        # let anything else still print.
        import sys

        exc = sys.exc_info()[1]
        if isinstance(exc, (BrokenPipeError, ConnectionResetError)):
            return
        super().handle_error(request, client_address)


@contextmanager
def serve(routes: dict[str, Route]) -> Iterator[_RouteServer]:
    """Serves `routes` (path -> `Route`) on an ephemeral `127.0.0.1` port for
    the lifetime of the `with` block. Usage::

        with serve({"/page": Route(body=b"<html>...</html>")}) as server:
            document = HtmlConnector().read(f"{server.base_url}/page")
    """
    server = _RouteServer(routes)
    thread = threading.Thread(target=server.serve_forever, daemon=True)
    thread.start()
    try:
        yield server
    finally:
        server.shutdown()
        thread.join(timeout=5)
        server.server_close()

File: sdk/typescript-langchain/tests/httpd.ts
/**
 * Minimal HTTP server harness for `HtmlConnector`'s URL-fetch tests (issue
 * #349) — a `node:http` server serving canned, in-memory responses,
 * synthesized at test time rather than depending on any external service
 * or recorded fixture. The mechanical mirror of the Python
 * `tests/_httpd.py` helper (issue #415).
 */

import { createServer, type IncomingMessage, type Server, type ServerResponse } from "node:http";

/**
 * One canned response. `location` set makes this a 302 redirect
 * (`status`/`contentType`/`body` are then ignored); otherwise `status` is
 * served with `body`, and `contentType` becomes the `Content-Type` header
 * (`null` sends no `Content-Type` header at all; the default is
 * `"text/html; charset=utf-8"`). `delay` (seconds) sleeps before
 * responding at all — enough to exercise a caller's per-phase `timeout`.
 * `chunkDelay` instead trickles `body` out in small pieces, sleeping
 * between each — every individual gap stays under a per-phase `timeout`,
 * so this is what exercises a caller's separate total-time budget
 * instead.
 */
export interface Route {
  status?: number;
  contentType?: string | null;
  body?: Buffer | Uint8Array;
  delay?: number;
  chunkDelay?: number;
  location?: string;
  headers?: Record<string, string>;
}

export interface RouteServer {
  readonly baseUrl: string;
  close(): Promise<void>;
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function handleRequest(
  req: IncomingMessage,
  res: ServerResponse,
  routes: Readonly<Record<string, Route>>,
): Promise<void> {
  res.on("error", () => {});
  req.on("error", () => {});

  // `req.url` is the full request target, query string included — route
  // registration is by path only, so a request carrying a query string
  // (the credential/token-stripping tests deliberately send one) must not
  // be matched against it too.
  const pathname = new URL(req.url ?? "/", "http://localhost").pathname;
  const route = routes[pathname];
  if (route === undefined) {
    res.writeHead(404);
    res.end();
    return;
  }

  if (route.delay) {
    await sleep(route.delay * 1000);
  }
  if (res.destroyed) {
    return;
  }

  if (route.location !== undefined) {
    res.writeHead(302, { Location: route.location });
    res.end();
    return;
  }

  const status = route.status ?? 200;
  const contentType = route.contentType === undefined ? "text/html; charset=utf-8" : route.contentType;
  const body = route.body ?? Buffer.alloc(0);
  const headers: Record<string, string> = { ...route.headers };
  if (contentType !== null) {
    headers["Content-Type"] = contentType;
  }

  if (route.chunkDelay) {
    res.writeHead(status, headers);
    const chunkSize = 16;
    for (let offset = 0; offset < body.length; offset += chunkSize) {
      if (res.destroyed) {
        return;
      }
      try {
        res.write(body.subarray(offset, offset + chunkSize));
      } catch {
        return;
      }
      await sleep(route.chunkDelay * 1000);
    }
    if (!res.destroyed) {
      res.end();
    }
    return;
  }

  headers["Content-Length"] = String(body.length);
  res.writeHead(status, headers);
  res.end(body);
}

function createRouteServer(routes: Readonly<Record<string, Route>>): Server {
  const server = createServer((req, res) => {
    void handleRequest(req, res, routes);
  });
  server.on("clientError", (_error, socket) => {
    if (!socket.destroyed) {
      socket.destroy();
    }
  });
  return server;
}

/**
 * Serves `routes` (path -> `Route`) on an ephemeral `127.0.0.1` port until
 * `close()` is called. Usage::
 *
 *     const server = await serve({ "/page": { body: Buffer.from("<html>...</html>") } });
 *     try {
 *       const document = await new HtmlConnector().read(`${server.baseUrl}/page`);
 *       ...
 *     } finally {
 *       await server.close();
 *     }
 */
export async function serve(routes: Readonly<Record<string, Route>>): Promise<RouteServer> {
  const server = createRouteServer(routes);
  await new Promise<void>((resolve) => {
    server.listen(0, "127.0.0.1", resolve);
  });
  const address = server.address();
  if (address === null || typeof address === "string") {
    throw new Error("failed to determine the test server's address");
  }
  const baseUrl = `http://127.0.0.1:${address.port}`;
  return {
    baseUrl,
    close: () =>
      new Promise<void>((resolve) => {
        server.close(() => resolve());
        // A client that abandons a response mid-stream (the timeout/
        // oversized-body tests deliberately do this) can leave a
        // keep-alive socket open; force-close everything so `close()`
        // doesn't hang waiting for it to go idle.
        server.closeAllConnections();
      }),
  };
}
